### ISSUE TITLE
Have @particle_input work with tuples of particles

### ISSUE THREAD

**Issue by namurphy** [PlasmaPy/PlasmaPy#336]
Presently the @particle_input decorator works by converting an argument annotated with the `Particle` class into a corresponding `Particle` instance.  An example:
```Python
from plasmapy.atomic import Particle, particle_input

@particle_input
def some_function(particle: Particle):
    print(particle.integer_charge)

>>> some_function("p+")
1
```
However, there are functions that take in tuples or lists of particles.  For example, `Coulomb_logarithm` has a `particles` argument that is for a tuple or list of the test particle and the target particle.  This usage is not supported by `@particle_input`, and needs to be implemented.  An example use case would be:

```Python
def another_function(particles: (Particle, Particle)):
    print(particle[0].integer_charge)
    print(particle[1].integer_charge)

>>> another_function(particles=('e-', 'p+'))
-1, 1
```
It should work for arguments that are tuples or lists.  Once this capability is added to `@particle_input`, then it will need to be implemented in `Coulomb_logarithm` and other functions in `collisions.py`.

**Comment by lemmatum** [comment]
Brainstorming on the design of this.... can we change all the particle inputs to be a tuple of particles and then add a check within the decorator to test if the correct number of particles (correct size tuple) has been passed to the function of interest?

you could then do something like
```Python
from plasmapy.atomic import Particle, particle_input

@particle_input(particleNum=1)
def some_function(particles):
    return
```

This will probably make generalizing from 1 particle to many particles a lot easier.

**Comment by ritiek** [comment]
It is now possible to pass collections of wannabe Particles to `@particle_input` with merge of #483. 

> can we change all the particle inputs to be a tuple of particles and then add a check within the decorator to test if the correct number of particles (correct size tuple) has been passed to the function of interest?

I am not sure if making it mandatory to pass a tuple of particles would be a good idea but thinking more about this maybe we could have something even better. Let's say we want to accept a collection containing 2 elements in `noblegases` and a collection containing 1 element in `radioactive`:
```python
>>> from plasmapy.atomic import Particle, particle_input

# Number of elements in a collection expected for each particle argument
>>> @particle_input(noblegases=2, radioactive=1)
>>> def some_function(noblegases: Particle, radioactive: Particle, blabla):
...     return

# `noblegases` requires a collection of 2 elements
# `radioactive` requires a collection of 1 element
>>> some_function(('He', 'Ne'), ('Th',), blabla='hello')
```
This would allow us to pass  2 or more wannabe Particles as separate arguments (`noblegases` and `radioactive` in above example) to a function decorated by `@particle_input` and having same argument name in `@particle_input` as in `some_function` should make it a bit intuitive.

**Comment by StanczakDominik** [comment]
That looks like the best solution here. Wouldn't it also make sense to
accept non tuple arguments with n=1?

**Comment by ritiek** [comment]
@StanczakDominik I don't feel it is needed. I am thinking about any argument that gets mentioned in `@particle_input` expects a collection of Particles, irrespective of n=1.

Maybe could keep the current functionality as is and add this functionality on top of it? One advantage of this could be to maintain readability depending on number of input elements in a collection. For example, if a function expects a lot of Particles in a collection, say 5, it would be more easier to express this with:
```python
>>> @particle_input(particles=5)
>>> def take_particles(particles: Particle):
...      return
```
instead of:
```python
>>> @particle_input
>>> def take_particles(particles: (Particle, Particle, Particle, Particle, Particle):
...     return
```

(but that does not mean 2nd code piece of code wouldn't work, it is user's choice on which method of representation they want to use)

Just my $0.02.

**Comment by namurphy** [comment]
It would be worth considering future applications where we might need expanded functionality, which would guide the direction that `particle_input` ends up taking.  Do we have any applications in mind where we would need multiple particles as inputs?

The application that motivated this issue was having to have a tuple of two charged particles as inputs to get the `Coulomb_logarithm` and other collision parameters.  Eventually we might end up having functions related to three-body recombination.  In #404, `IonizationStates` can accept a `list` of particle arguments or a `dict` containing ionization state information, but the code to deal with that flexible input is pretty specialized and I don't think it would be worth putting it into a general purpose decorator.  Are there other situations that we should be thinking of?

**Comment by namurphy** [comment]
Closed by #483.
